osrf/vrx
# Navigation scoring plugin port

Port of the navigation scoring plugin to new Gazebosim / ros2. 

## Requirements
* Gazebosim must be installed from source.
* Gamepad is needed for WAMV teleop.

## Preparation
The following adjustments must be made to the individual repositories within the gazebo workspace before building:

### gz-fuel-tools
* Make sure `gz-fuel-tools` is checked out to commit `7b6ce8a4ab4cbfd724eefba7f2e57892671abfc0`
  ```
  cd gazebo_ws/src/gz-fuel-tools
  git checkout 7b6ce8a4ab4cbfd724eefba7f2e57892671abfc0
  ```
* Previous commits may work, but the 2 subsequent commits have a bug that will cause a runtime crash.

### enable teleoperation
* We need to teleoperate the WAMV to make sure the gates are doing the right thing. This requires functionality that has not yet been merged from v6 to v7, so we need to use the `6-to-7` branch:
  ```
  cd gazebo_ws/src/gz-sim
  git checkout 6-to-7-20221021
  ```
### rebuild
After making the above changes, rebuild gazebo:
```
cd ~/gazebo_ws
colcon build --merge-install
```
Some warnings are expected. Building a second time should produce no stderr output.

## Build
In a separate terminal, checkout the `M1chaelM/navigation_scoring_plugin` branch and build it:
```
cd ~/vrx_ws/src/vrx
git checkout M1chaelM/navigation_scoring_plugin
cd ~/vrx_ws
GZ_VERSION=garden colcon build --merge-install
```
Again, some warnings are expected for the first build.

## Run
Open three new terminals. In each, source the Gazebo and VRX workspaces:
```
cd vrx_ws
source ~/gazebo_ws/install/setup.bash
. install/setup.bash
```
Run the simulation in one terminal:
```
ros2 launch vrx_gz competition.launch.py world:=navigation_task
```
Run teleop in a second terminal:
```
ros2 launch vrx_gz usv_joy_teleop.py
```
Echo the `task/info` topic in the third terminal:
```
ros2 topic echo /vrx/task/info
```

## Test
* Wait for the task to enter `onrunning` state.
* Use the gamepad to drive the WAMV through the gates in the course.
* Check that a "New gate crossed!" message is printed to terminal each time a gate is crossed.
* Check that the simulation exits after all 3 gates are crossed.
* Check that the score printed in the `vrx/task/info` message is reasonable. It should be equal to the elapsed time between the crossing the entrance gate the exit gate.
* Check that crossing a gate out of order terminates the task.
* Check that crossing a gate in the wrong direction terminates the task.

## Known issues
* The WAMV does not lock, so it is possible to drive through the first gate before the simulation enters the `onrunning` state. This will cause the gate to be ignored. Passing a second gate will then be treated as crossing a gate out of order, which terminates the simulation.
* Currently the entire simulation terminates the moment the task is over. In Gazebo Classic, my memory is that it stayed up but  just froze in place.
* If we use this task again we may want to revisit how we calculate the score. Starting the timer at the entrance gate seems odd.
